### PR TITLE
ParameterSettings properly implemented

### LINKED ISSUE #256: Revit_Toolkit: Parameter Mapping for Structural classes

### LINKED ISSUE #451: Revit_Toolkit: Replace hard coded "BHE_Material Grade" parameter name
#### Description:
At the moment the material grade is based on the hard coded parameter with name _BHE_Material Grade_. This needs to be made more flexible, preferably with an implementation of `ParameterMap`.

### LINKED ISSUE #461: Revit_Toolkit: UpdateBuildingElementTypeByCustomData is a temporary fix
#### Description:
Following the code comment: `BH.UI.Revit.Engine.Query.UpdateBuildingElementTypeByCustomData` method is a temp fix - should be resolved.

### LINKED ISSUE #482: Revit_Toolkit: Refactor MapSettings
#### Description:
Current implementation of parameter mapping using `MapSettings` is not fully functional: it is quite limited + used only in few cases (an in many not). It would be good to implement it as follows:
- create an external file (.json?) that would contain information which BHoM properties should be coupled with which Revit parameters etc.
- create a method that loads the above file as well as other files that contain project/user specific data in same format

This is likely to supersede #256, #258, #451.

### LINKED ISSUE #533: Revit_Toolkit: Merge TagsParameterName into MapSettings
#### Description:

`TagsParameterName` is currently a separate, floating string value - it should become a part of `MapSettings`.

### LINKED ISSUE #563: Is UpdateValues method needed at all?
#### Description

`BH.UI.Revit.Engine.Modify.UpdateValues` method is used on ToBHoM convert methods. Not sure what it does as it is used mainly by BEnv. Couldn't we either get rid of it or merge into `SetCustomData`?

### LINKED ISSUE #725: Revit_Toolkit: Information duplicated on BEnv converts
#### Description:

`EnvironmentPanelFromRevit` and `OpeningFromRevit` copy over the underlying type parameters to BHoM object's `CustomData`, even though they are also copied to `CustomData` of this same object's `Construction`.

https://github.com/BHoM/Revit_Toolkit/blob/7066dfab9a4bedf8ac000dfdb241c4ff7080d49c/Engine_Revit_UI/Convert/Environment/FromRevit/Panel.cs#L497-L498

https://github.com/BHoM/Revit_Toolkit/blob/7066dfab9a4bedf8ac000dfdb241c4ff7080d49c/Engine_Revit_UI/Convert/Environment/FromRevit/Opening.cs#L102-L103

@FraserGreenroyd is that necessary?

### PR DESCRIPTION
**Pull request by pawelbaran**
### NOTE: Depends on 

https://github.com/BHoM/BHoM_Datasets/pull/47

   
### Issues addressed by this PR

Closes #256
Closes #451
Closes #461
Closes #482
Closes #533
Closes #563
Closes #725

### Test files

PR-specific test on [SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2FRevit%5FToolkit%2DIssues256%2C451%2C459%2C482%2C533&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4)
Macro pull tests on [SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2F%5FMacro%2FPull&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4)

### Changelog

- `MapSettings` has been refactored into `ParameterSettings` type that has a `ParameterMaps` property, where each ParameterMap is related to a certain type and has a collection of `ParameterLink` objects that define the relationship between type property (or `CustomData` key) and Revit parameter names (OR names of Revit element's type parameter names - then the parameter name in `ParameterLink` needs to be preceded with a prefix _type:_).

image

- Necessary supplementary `Create`, `Modify` and `Query` methods were added to support the above.
- `SetIdentifiers`, `SetCustomData` and `SetProperties` (previously `UpdateValues)` have been unified to work equally in all `FromRevit` converts.
- Two properties related to variables used by the Revit_Toolkit were added to `ParameterSettings`:
    - `TagsParameter` that identifies the parameter to write or read `Tags`
    - `MaterialGradeParameter` that identifies the parameter that contains information about material class
- Related code base has been cleaned up.

### Additional comments

Although the number of changed files is massive, most of the changes are either cosmetical (method renaming, code cleanup etc.) or testable greenfield (see the test file), so I hope it will be reasonably easy to review. Still, however, utmost caution is recommended (particularly on BEnv side @FraserGreenroyd @MajaLindroth).

Each `FromRevit` method (apart from material converts - these will be handled in the next wave) should contain following methods:

```
 //Set identifiers, parameters & custom data
result.SetIdentifiers(element);
result.SetCustomData(element, settings.ParameterSettings);
result.SetProperties(element, settings.ParameterSettings);
```

This PR will impose breaking changes due to complete refactor of `MapSettings` into `ParameterSettings`. This, however, is unfortunately impossible to avoid due to the scope of changes (in this PR and coming ones). @al-fisher @rwemay FYI

And finally a framework question: atm this PR works under assumption that the only class from which BHoM types can inherit is `BHoMObject` (i.e. besides `BHoMObject` the types are not allowed from any other non-interface type). Is that a correct assumption @IsakNaslundBh @al-fisher @FraserGreenroyd?

### CODE REVIEW

**Comment by pawelbaran** [comment]
@MajaLindroth could you give some more info on what did go wrong in the XML conversion? I am not familiar with the process so will be hard for me to trace the issue. Thanks!

**Comment by FraserGreenroyd** [comment]
@pawelbaran when pulling the model none of the shading elements were correctly attributed as shading elements (which was then seen when pushing to XML), but if you:

 - Pull an energy analysis request model
 - Filter all objects to just the environment panels
 - Filter environment panels by type for shade
 - You should have a few, but currently there are none.

It is the same for some curtain walls and other types as well.

**Comment by pawelbaran** [comment]
Oke @FraserGreenroyd @MajaLindroth the issue comes from me resolving #725 and making #461 even worse. I will try to fix it once and for all now.

File: Adapter_Revit_UI/CRUD/Read.cs [hunk 1]
```diff
@@ -191,7 +186,7 @@ public static IEnumerable<IBHoMObject> Read(Element element, Discipline discipli
             //Assign Tags
             string tagsParameterName = null;
             if (settings != null)
-                tagsParameterName = settings.TagsParameterName;
+                tagsParameterName = settings.ParameterSettings.TagsParameter;
```
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Adapter_Revit_UI/CRUD/Read.cs:189]: I have refined the concept and stopped using _Name_ suffix everywhere in `ParameterSettings`. I've found it obvious that we are talking about parameter names and wanted to shorten the properties for UX reasons. However, happy to put it under wider discussion if you have a strong opinion on that one

**Comment by pawelbaran** [comment]
A few updates to this PR:
- setting `enum` properties of BHoM objects based on Revit parameters has been enabled with some degree of flexibility (number of words in param value == number of capitals in `enum` && all words need to be contained inside `enum` case insensitive)
- setting properties of BHoM objects based on parameters of Revit element types has been enabled
- #461 has been resolved based on the two above

PR description as well as test files have been updated to reflect the changes.

@MajaLindroth @FraserGreenroyd ready for another review.

**Comment by pawelbaran** [comment]
@FraserGreenroyd I have just looked into the problem, with the following findings:
- On this PR the arrowed panels come out as `PanelType.Wall` or `PanelType.Floor`
- On master they come out as `PanelType.Undefined`
- There is no `Air` in `BH.oM.Environment.Elements.PanelType`

Cannot think of any explanation for that, as long as you are not using any `CustomData` downstream - then I would guess you agreed to delete one item too much in #725.

**Comment by pawelbaran** [comment]
Looks like we are good to go for another round of testing @MajaLindroth @FraserGreenroyd

**Pull request merged**